backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java

```diff
@@ -2247,4 +2250,55 @@ public void deleteLogVersionedInputs(Session session, Long repoId, String commit
           .build();
     }
   }
+
+  @Override
+  public Entry<String, String> getExperimentRunArtifactsS3PathAndMultipartUploadID(
+      String experimentRunId, String key, long partNumber, S3KeyFunction initializeMultipart)
+      throws ModelDBException {
+    try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
+      ExperimentRunEntity experimentRunObj =
+          session.get(ExperimentRunEntity.class, experimentRunId);
+      if (experimentRunObj == null) {
+        LOGGER.warn(ModelDBMessages.EXP_RUN_NOT_FOUND_ERROR_MSG);
+        Status status =
+            Status.newBuilder()
+                .setCode(Code.NOT_FOUND_VALUE)
+                .setMessage(ModelDBMessages.EXP_RUN_NOT_FOUND_ERROR_MSG)
+                .build();
+        throw StatusProto.toStatusRuntimeException(status);
+      }
+      Map<String, List<ArtifactEntity>> artifactEntityMap = experimentRunObj.getArtifactEntityMap();
+
+      List<ArtifactEntity> result =
+          (artifactEntityMap != null && artifactEntityMap.containsKey(ModelDBConstants.ARTIFACTS))
+              ? artifactEntityMap.get(ModelDBConstants.ARTIFACTS)
+              : Collections.emptyList();
+      for (ArtifactEntity artifactEntity : result) {
+        if (artifactEntity.getKey().equals(key)) {
+          String uploadId;
+          if (partNumber != 0) {
+            uploadId = artifactEntity.getUploadId();
+            if (uploadId == null) {
+              uploadId = initializeMultipart.apply(key).orElse(null);
```

Comment:
`initializeMultipart()` needs the S3 key (which is `path` from the artifact), not the ModelDB artifact key!
```suggestion
              uploadId = initializeMultipart.apply(artifactEntity.getPath()).orElse(null);
```
This is why we get `NoSuchUpload` from S3 when we try to use the presigned URL:
The key used to generate the URL was different from the one used for `initializeMultipart()`.

I have verified that this change fixes the `NoSuchUpload` error.
If `initializeMultipart()` is called anywhere else, make sure those are correct as well!

Reply:
Thank you very much! That works